sdk/core/core/internal/az_log_internal.h

```diff
@@ -12,9 +12,23 @@
 #include <_az_cfg_prefix.h>
 
 // If the user hasn't registered any classifications, then we log everything.
-bool _az_log_should_write(az_log_classification classification);
 
+#ifndef AZ_NO_LOGGING
+bool _az_log_should_write(az_log_classification classification);
 void _az_log_write(az_log_classification classification, az_span message);
+#else
+AZ_INLINE bool _az_log_should_write(az_log_classification classification)
+{
+  (void)classification;
+  return false;
+}
+
+AZ_INLINE void _az_log_write(az_log_classification classification, az_span message)
+{
+  (void)classification;
+  (void)message;
+}
+#endif // AZ_NO_LOGGING
```

Comment:
any reason not to use same pattern as used by AZ_NO_PRECONDITION_CHECKING?

#ifdef AZ_NO_PRECONDITION_CHECKING
#define _az_PRECONDITION(condition)
#else
#define _az_PRECONDITION(condition) \
  do \
  { \
    if (!(condition)) \
    { \
      az_precondition_failed_get_callback()(); \
    } \
  } while (0)
#endif // AZ_NO_PRECONDITION_CHECKING



Also, current implementation can you confirm that there is no code in binary for  _az_log_write, when using no logging,

Reply:
Tagging @TiejunMS for code review

Reply:
I'm OK with the inline function. The compiler is smart enough to strip out the log functions when `AZ_NO_LOGGING` is defined. Here are the results of experiments I did.
1. Keep source code of convenience layer (CL) unchanged and disable logging in CL. RO code is 8000 and RO data is 303. [Link](https://expresslogic.visualstudio.com/X-Ware/_build/results?buildId=14313&view=logs&j=84d3beb9-ba6d-5951-beee-b2a2468262b8&t=9c5b81df-f200-5a06-d6b4-ebad7c262ef0&l=46).
1. Use `_az_log_write` in CL with az_log **enabled**. RO code is 10052 and RO data is 3903. [Link](https://expresslogic.visualstudio.com/X-Ware/_build/results?buildId=14314&view=logs&j=84d3beb9-ba6d-5951-beee-b2a2468262b8&t=9c5b81df-f200-5a06-d6b4-ebad7c262ef0&l=46).
1.  Use `_az_log_write` in CL with az_log **disabled**. RO code is 7912 and RO data is 303. [Link](https://expresslogic.visualstudio.com/X-Ware/_build/results?buildId=14317&view=logs&j=84d3beb9-ba6d-5951-beee-b2a2468262b8&t=9c5b81df-f200-5a06-d6b4-ebad7c262ef0&l=45).

Reply:
But if customer wants to debug a debug build and the binary is not fitting (because of logs and no optimization). In that case customer can try AZ_NO_LOGGING to reduce the binary size and see if binary fits in the device (because literal will be removed), but with current change it looks like this is not possible [Link](https://expresslogic.visualstudio.com/X-Ware/_build/results?buildId=14313&view=logs&j=84d3beb9-ba6d-5951-beee-b2a2468262b8&t=0e610488-0db1-5249-3946-6cea4ec1597a), as literal are still pinned in the binary.

Reply:
You raise a good point! I didn't think about it. In this case, it is worth converting az_log_write to macros.

Reply:
@TiejunMS, could you please try to see what numbers you can get from this branch with the optimization flags (https://www.rapidtables.com/code/linux/gcc/gcc-o.html - I played on https://godbolt.org/, it looks like any `-O` would be sufficient - either just `-O`, or `-O3`) and also what numbers you can get with the branch `antk/preconditions-inline-fns` (https://github.com/Azure/azure-sdk-for-c/pull/714).

Reply:
@antkmsft , we don't have project for ARM GCC. So I didn't check the code size generated by it. For `antk/preconditions-inline-fns` branch, here is [link](https://expresslogic.visualstudio.com/X-Ware/_build/results?buildId=14597&view=logs&j=84d3beb9-ba6d-5951-beee-b2a2468262b8&t=9c5b81df-f200-5a06-d6b4-ebad7c262ef0&l=46) to map file.

Reply:
@TiejunMS what steps do you take to generate those, what prerequisites do you need, and how long does it take? It would be great if you could share those so we could run it as well, especially based on local changes.

Reply:
@ahsonkhan , it is the build pipeline of convenience layer development. The agent for IAR build is run on a physical Windows machine with IAR installed. This [pipeline](https://expresslogic.visualstudio.com/X-Ware/_build?definitionId=139&_a=summary) is triggered by PR and CI. And we can queue the build manually also. It takes 5 minutes from checkout till test is done. 

SDK repository is a submodule of [convenience layer repository](https://expresslogic.visualstudio.com/X-Ware/_git/azure-services-port). If you need to do any tests by this pipeline, you need to update the submodule to your local branch and push to DevOps. Then manually queue a build in this pipeline.

Reply:
@JeffreyRichter , in case your compiler is not very good, and/or if you don't use optimization flags, function calls and string literals are not being removed. Should we go with the macros? (https://github.com/Azure/azure-sdk-for-c/pull/710). They don't rely on the compiler optimizing, and ensure that the code is not included under any circumstances.

I propose to merge https://github.com/Azure/azure-sdk-for-c/pull/710 here, and close https://github.com/Azure/azure-sdk-for-c/pull/714.

Reply:
Jeffrey agreed to use macros, now I need someone's approval for the https://github.com/Azure/azure-sdk-for-c/pull/710 please. It won't go to `master`, it will go to this branch (the rules are too strict).

Reply:
@hihigupt , we use macros now.